### PR TITLE
Assignment 4

### PR DESCRIPTION
**Pull request by Cyc10n3z**

### CODE REVIEW

File: source/ControllableHero.hx
Comment (SamBumgardner) [inline source/ControllableHero.hx]: this is fine, though it still has the problem of tying our game logic to the framerate (which isn't a problem for this assignment, but something we want to improve in the future).

If you want to use the built-in velocity handling from `FlxObject`, you could instead do something like
```
velocity.set(0,0);
if (FlxG.keys.anyPressed([UP, W]))
{
    velocity.y -= speed;
}
...
```
Every time update is called, velocity is set back to 0, then x and y velocity is changed according to which inputs are pressed. This does mean that a player moving diagonally covers more total distance (10 vs 10 * sqrt(2)) than just moving in a cardinal direction, but it does effectively separate our game logic from the visual framerate.

File: source/ScreensaverHero.hx
Comment (SamBumgardner) [inline source/ScreensaverHero.hx:19]: Why did you choose to include `true` as an argument to this method?

File: source/ScreensaverHero.hx
Comment (SamBumgardner) [inline source/ScreensaverHero.hx]: The goal of the assignment is to make it possible for the color to change as the ScreensaverHero moves across the screen:

> Inside the ScreensaverHero's update() function, set its [color property](https://api.haxeflixel.com/flixel/FlxSprite.html#color) to a value that is somehow determined by screen position (how is up to you).

You are setting the color property inside of the update function, so we're halfway there. Now you need to make the value you're putting into `color` be determined by your screen position in some way.

It could be as simple as "if it's on the left side of the screen, it's blue, if it's on the right side it's green" or you could use some method like `FlxColor.fromRGB()` to generate a color, using the `x` and `y` properties of your `ScreensaverHero` to create numbers you pass into the method.

We'll need to get this done before we can merge.

File: source/ControllableHero.hx
Comment (SamBumgardner) [inline source/ControllableHero.hx]: Creating your own variable to track the speed this object should move at is totally reasonable. Might be good to specify the type (int or float) to be absolutely clear. You can also add [access modifiers](https://haxe.org/manual/class-field-property.html) if you want to make this value totally unchangeable if you expect it to be a constant.

File: source/ControllableHero.hx
Comment (SamBumgardner) [inline source/ControllableHero.hx]: these lines are unnecessary. `x` and `y` are properties defined in the `FlxObject` class that this (indirectly) inherits from - when you call `super(x, y)`, it's executing our parent class's logic, which is what ultimately ends up setting our `x` and `y` initial values.

Also, the `this` keyword is generally not necessary - the only `x` and `y` variables that are available in this method (since you haven't declared any local variables) are the instance variables attached to this class - `this.x` and `x` are both just referencing the same thing.

File: source/ControllableHero.hx
Comment (SamBumgardner) [inline source/ControllableHero.hx]: The 
```
var myRectangle = new FlxSprite(x, y);
```
Isn't impacting the game in a useful way.
Remember that `FlxSprite` class (and everything that inherits from it) is like a "character" or "visible, moveable object" in our game world - because this class inherits from `FlxSprite`, it's already going to have all of the methods and fields of its parent. It has the ability to render graphics and get drawn to the screen.

There's nothing that stops you from creating a new FlxSprite object here, but I don't think it ends up making much sense. Our "character" is the `ControllableHero` class, and your call to `makeGraphic` is what sets up its lovely orange rectangle. 

The `myRectangle` that you create here is a newly made FlxSprite, with all of the same underlying capabilites and fields as this class. If it were added to the playstate somehow, then you'd see it's little graphic on the screen too - you can think of each FlxSprite as its own separate entity, so it *usually* doesn't make sense for one thing in the game to create another.

The `makeGraphic` call is good, but you don't have to hard-code that `0xFFFFA500` value into the method call - instead you can just put in `FlxColor.ORANGE` and retrieve the value that FlxColor provides statically for anyone to use. The code would end up looking like
```
makeGraphic(30, 10, FlxColor.ORANGE, true);
```
(and you'd have this import added to the top of your file)
```
import flixel.util.FlxColor;
```
Using named variables like this definitely improves readability of the code without you having to write extra comments to explain what's going on. It's nice that you have the comment, since the hard-coded value would be much harder to interpret otherwise, but it's even better to not need it in the first place.

File: source/PlayState.hx
Comment (SamBumgardner) [inline source/PlayState.hx]: If you're thinking of finishing this up during this PR, let's do it!

If you don't, then let's delete this leftover commented-out stuff so our `main` branch stays nice and clean.

Same applies to the commented lines inside `update()` below.

**Review by SamBumgardner — CHANGES_REQUESTED** [review]
I have a variety of comments, but the blocking issue that needs to be addressed before merge is making the `ScreensaverHero` change color as it moves around the screen.

Other stuff could be worth doing too, but I leave those changes up to you.

File: source/ScreensaverHero.hx
Comment (Cyc10n3z) [reply source/ScreensaverHero.hx:19]: When I was reading about makeGraphic() the 'unique parameter' (boolean value) allows the graphic to be a unique instance in the graphics cache, which allows you to modify the pixels of a sprite without modifying other sprites with the same "bitmap data". I am not sure I used that correctly or for this assignment it is unnecessary?

Comment (SamBumgardner) [reply source/ScreensaverHero.hx:19]: Gotcha, that makes sense.

You're correct that making the graphic unique protects the instances from accidentally modifying the underlying bitmap data, but setting `color` isn't actually changing that so `unique` ends up being unnecessary.

The magic of graphics pipelines is mostly outside of the scope of our class, but you can think of the `colorTransform()` method as a sort of bonus filter that Haxeflixel applies over that specific `ScreensaverHero` when it's time to draw it to the screen.

It's possible to do other things that actually modify the underlying graphic (in which case keeping them unique is probably desirable) but the color tinting isn't one of them.

**Review by SamBumgardner — APPROVED** [review]
Nice work on the changes. Essential stuff is complete, so you can merge when ready. If you end up making more changes, let me know and I'll come back for re-review.

**Pull request merged**
